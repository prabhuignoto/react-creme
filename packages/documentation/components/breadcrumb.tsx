import React from 'react';
import { BreadCrumb } from '../../lib/components';
function breadcrumb() {
  return (
    <div>
      <div>
        <BreadCrumb icon="arrow">
          <span>one</span>
          <span>two</span>
          <span>three</span>
        </BreadCrumb>
      </div>
      <div>
        <BreadCrumb size="md">
          <span>one</span>
          <span>two</span>
          <span>three</span>
        </BreadCrumb>
      </div>
      <div>
        <BreadCrumb icon="slash" size="lg">
          <span>one</span>
          <span>two</span>
          <span>three</span>
        </BreadCrumb>
      </div>
    </div>
  );
}

export default breadcrumb;
