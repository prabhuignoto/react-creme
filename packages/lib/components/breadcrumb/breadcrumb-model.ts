import React, { ReactNode } from 'react';

export interface BreadCrumbProps extends BreadCrumbCommonProps {
  children: React.ReactNode;
  onSelected?: (selectedIndex?: string) => void;
  selectedCrumbIndex?: number;
}

export interface BreadCrumbItemProps extends BreadCrumbCommonProps {
  child: ReactNode;
  id: string;
  index: number;
  onSelected?: (id: string, index: number) => void;
  selected?: boolean;
  showChevron?: boolean;
}

export interface BreadCrumbCommonProps {
  icon?: 'chevron' | 'arrow' | 'slash';
  size?: 'sm' | 'md' | 'lg';
}
