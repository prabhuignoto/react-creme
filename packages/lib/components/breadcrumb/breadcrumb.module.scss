@use '../../design/core.scss';
@use '../../design/list.scss';
@use '../../design/theme.scss';
@use '../../design/animate.scss';
@use '../../design/font.scss';

$sizes: (sm, md, lg);

.bread_crumbs_wrapper {
  @extend %left;
  margin: 0;
  padding: 0;
}

.bread_crumb {
  @extend %list-item;
  cursor: pointer;
}

.bread_crumb_icon {
  margin-left: 0.5rem;

  @each $size in $sizes {
    &.bread_crumb_icon_#{$size} {
      @extend %icon-#{$size};
    }
  }

  svg {
    color: theme.$primary;
  }
}

.bread_crumb_node {
  @extend %center;
  @extend %border-radius;
  padding: 0.25rem;

  @each $size in $sizes {
    &.bread_crumb_node_#{$size} {
      > * {
        @extend %font-#{$size};
      }
    }
  }
}

.slash {
  transform: rotate(120deg);
}
