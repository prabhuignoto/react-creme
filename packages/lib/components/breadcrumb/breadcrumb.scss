@use '../../design/layout.scss';
@use '../../design/list.scss';
@use '../../design/icon.scss';
@use '../../design/colors.scss';
@use '../../design/animate.scss';
@use '../../design/font.scss';

$sizes: (sm, md, lg);

.rc-bread-crumbs-wrapper {
  @extend %left;
  margin: 0;
  padding: 0;
}

.rc-bread-crumb {
  @extend %list-item;
  cursor: pointer;
}

.rc-bread-crumb-icon {
  margin-left: 0.5rem;

  @each $size in $sizes {
    &.rc-bread-crumb-icon-#{$size} {
      @extend %icon-#{$size};
    }
  }

  svg {
    color: colors.$primary;
  }
}

.rc-bread-crumb-node {
  @extend %center;
  @extend %border-radius;
  //
  padding: 0.25rem;

  &:hover {
    // background-color: rgba(colors.$secondary-rgb, 0.25);
    // color: colors.$tertiary;
  }

  &.rc-bread-crumb-node-selected {
    background-color: rgba(colors.$secondary-rgb, 0.5);
    color: colors.$tertiary;
  }

  @each $size in $sizes {
    &.rc-bread-crumb-node-#{$size} {
      > * {
        @extend %font-#{$size};
      }
    }
  }
}
