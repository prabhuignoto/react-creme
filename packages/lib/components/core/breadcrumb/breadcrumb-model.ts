import { ReactNode } from 'react';

export type BreadCrumbProps = {
  RTL?: boolean;
  children?: ReactNode | ReactNode[];
  focusable?: boolean;
  icon?: 'chevron' | 'arrow' | 'slash';
  links: string[];
  onSelected?: (selected?: string) => void;
  selectedCrumbIndex?: number;
  size?: 'sm' | 'md' | 'lg';
};

export type BreadCrumbItemProps = Pick<
  BreadCrumbProps,
  'icon' | 'selectedCrumbIndex' | 'size' | 'focusable' | 'RTL'
> & {
  child: ReactNode;
  id: string;
  index: number;
  name: string;
  onSelected?: (id: string, name: string) => void;
  selected?: boolean;
  showChevron?: boolean;
};
