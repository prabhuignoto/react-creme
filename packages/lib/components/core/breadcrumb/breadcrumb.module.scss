@use '../../../design/core.scss';
@use '../../../design/list.scss';
@use '../../../design/theme.scss';
@use '../../../design/animate.scss';
@use '../../../design/font.scss';

$sizes: (sm, md, lg);

.bread_crumbs_wrapper {
  @extend %left;
  margin: 0;
  padding: 0;

  &.rtl {
    direction: rtl;
  }
}

.bread_crumb {
  @extend %list-item;
  cursor: pointer;
}

.bread_crumb_icon {
  &:not(.rtl_icon) {
    margin-left: 0.5rem;
  }

  &.rtl_icon {
    margin-right: 0.5rem;
    transform: rotate(-180deg);
  }

  @each $size in $sizes {
    &.bread_crumb_icon_#{$size} {
      @extend %icon-#{$size};
    }
  }

  svg {
    color: theme.$primary;
  }
}

.bread_crumb_node {
  @extend %center;
  @extend %border-radius;
  padding: 0.25rem;

  @each $size in $sizes {
    &.bread_crumb_node_#{$size} {
      > * {
        @extend %font-#{$size};
      }
    }
  }
}

.slash {
  transform: rotate(120deg);
}
