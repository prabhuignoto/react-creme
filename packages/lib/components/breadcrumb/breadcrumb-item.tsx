import classNames from 'classnames';
import React, { FunctionComponent, useMemo, useRef } from 'react';
import { ArrowRightIcon, ChevronRightIcon } from '../../icons';
import useFocusNew from '../common/effects/useFocusNew';
import { BreadCrumbItemProps } from './breadcrumb-model';

const BreadCrumbItem: FunctionComponent<BreadCrumbItemProps> = React.memo(
  ({
    id,
    onClick,
    child,
    showChevron,
    icon = 'chevron',
    size = 'sm',
  }: BreadCrumbItemProps) => {
    const ref = useRef<HTMLSpanElement>(null);
    useFocusNew(ref);

    const breadCrumbIcon = useMemo(() => {
      return classNames('bread-crumb-icon', {
        [`bread-crumb-icon-${size}`]: true,
      });
    }, []);

    const breadCrumbNode = useMemo(() => {
      return classNames('bread-crumb-node', {
        [`bread-crumb-node-${size}`]: true,
      });
    }, []);

    return (
      <li className="bread-crumb" key={id}>
        <span
          className={breadCrumbNode}
          ref={ref}
          tabIndex={0}
          onClick={onClick}
        >
          {child}
        </span>
        {showChevron && (
          <span className={breadCrumbIcon}>
            {icon === 'chevron' && <ChevronRightIcon />}
            {icon === 'arrow' && <ArrowRightIcon />}
            {icon === 'slash' && '/'}
          </span>
        )}
      </li>
    );
  }
);

BreadCrumbItem.displayName = 'BreadCrumbItem';

export { BreadCrumbItem };
