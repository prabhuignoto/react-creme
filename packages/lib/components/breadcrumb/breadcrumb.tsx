import { nanoid } from 'nanoid';
import React, { useCallback, useState } from 'react';
import { BreadCrumbItem } from './breadcrumb-item';
import { BreadCrumbProps } from './breadcrumb-model';
import styles from './breadcrumb.module.scss';

interface BreadCrumbItemModel {
  id: string;
  name: string;
  selected: boolean;
}

const BreadCrumb: React.FunctionComponent<BreadCrumbProps> = ({
  children,
  onSelected,
  icon = 'chevron',
  size = 'sm',
  focusable = true,
  selectedCrumbIndex = 0,
  links = [],
}) => {
  const [items, setItems] = useState<BreadCrumbItemModel[]>(
    links.map((link, index) => ({
      id: nanoid(),
      name: link,
      selected: selectedCrumbIndex === index,
    }))
  );

  const handleSelection = useCallback((id, name) => {
    setItems(prev =>
      prev.map(item => ({
        ...item,
        selected: item.id === id,
      }))
    );

    onSelected?.(name);
  }, []);

  return (
    <ul
      className={styles.bread_crumbs_wrapper}
      role="navigation"
      aria-label="breadcrumbs"
    >
      {items.map(({ id, selected, name }, index) => (
        <BreadCrumbItem
          child={Array.isArray(children) && children[index]}
          id={id}
          onSelected={handleSelection}
          showChevron={index < items.length - 1}
          key={id}
          icon={icon}
          size={size}
          selected={selected}
          index={index}
          focusable={focusable}
          name={name}
        />
      ))}
    </ul>
  );
};

BreadCrumb.displayName = 'BreadCrumb';

export { BreadCrumb };
