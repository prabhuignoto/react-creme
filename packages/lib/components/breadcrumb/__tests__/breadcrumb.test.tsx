import { fireEvent, render, waitFor } from '@testing-library/react';
import { it, vi } from 'vitest';
import { BreadCrumb } from '../breadcrumb';

describe('Breadcrumb', () => {
  it('should render', async () => {
    const { getByRole, getByText } = render(
      <BreadCrumb links={['google', 'yahoo', 'amazon']}></BreadCrumb>
    );
    await waitFor(() => {
      expect(getByRole('navigation')).toBeInTheDocument();

      expect(getByText('google')).toBeInTheDocument();
      expect(getByText('yahoo')).toBeInTheDocument();
    });
  });

  it('should handler is called', async () => {
    const handler = vi.fn();
    const { getByText } = render(
      <BreadCrumb
        onSelected={handler}
        links={['google', 'yahoo', 'amazon']}
      ></BreadCrumb>
    );

    expect(getByText('google')).toBeInTheDocument();
    fireEvent.click(getByText('google'));

    expect(handler).toBeCalledWith('google');
  });

  it('should render snapshot', () => {
    const { getByRole } = render(
      <BreadCrumb links={['google', 'yahoo', 'amazon']}></BreadCrumb>,
      {
        container: document.body,
      }
    );

    expect(getByRole('navigation')).toMatchSnapshot();
  });
});
