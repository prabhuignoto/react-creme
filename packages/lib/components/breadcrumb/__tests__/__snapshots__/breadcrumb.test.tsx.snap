// Vitest Snapshot v1

exports[`Breadcrumb > should render snapshot 1`] = `
<ul
  aria-label="breadcrumbs"
  class="_bread_crumbs_wrapper_1ju3c_42"
  role="navigation"
>
  <li
    class="_bread_crumb_1ju3c_1"
  >
    <span
      class="_bread_crumb_node_1ju3c_1 _bread_crumb_node_sm_1ju3c_5"
    >
      <a
        class="_link_apynu_1 _sm_apynu_19 _highlight_apynu_82"
        style="outline: none; position: relative;"
        tabindex="0"
      >
        <span
          class="_text_apynu_19"
        >
          google
        </span>
        <span
          class="focus_ring"
          style="width: 6px; height: 6px;"
        />
      </a>
    </span>
    <span
      class="_bread_crumb_icon_1ju3c_17 _bread_crumb_icon_sm_1ju3c_17"
    >
      <svg
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 18l6-6-6-6"
        />
      </svg>
    </span>
  </li>
  <li
    class="_bread_crumb_1ju3c_1"
  >
    <span
      class="_bread_crumb_node_1ju3c_1 _bread_crumb_node_sm_1ju3c_5"
    >
      <a
        class="_link_apynu_1 _sm_apynu_19"
        style="outline: none; position: relative;"
        tabindex="0"
      >
        <span
          class="_text_apynu_19"
        >
          yahoo
        </span>
        <span
          class="focus_ring"
          style="width: 6px; height: 6px;"
        />
      </a>
    </span>
    <span
      class="_bread_crumb_icon_1ju3c_17 _bread_crumb_icon_sm_1ju3c_17"
    >
      <svg
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 18l6-6-6-6"
        />
      </svg>
    </span>
  </li>
  <li
    class="_bread_crumb_1ju3c_1"
  >
    <span
      class="_bread_crumb_node_1ju3c_1 _bread_crumb_node_sm_1ju3c_5"
    >
      <a
        class="_link_apynu_1 _sm_apynu_19"
        style="outline: none; position: relative;"
        tabindex="0"
      >
        <span
          class="_text_apynu_19"
        >
          amazon
        </span>
        <span
          class="focus_ring"
          style="width: 6px; height: 6px;"
        />
      </a>
    </span>
  </li>
</ul>
`;
